polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h

```diff
@@ -434,6 +432,56 @@ inline raw_ostream &operator<<(raw_ostream &os,
   return os;
 }
 
+class MemoryAccessAnalysis {
+public:
+  MemoryAccessAnalysis(Operation *op, AnalysisManager &am);
+
+  bool isInvalidated(const AnalysisManager::PreservedAnalyses &pa);
+
+  /// Returns the operation this analysis was constructed from.
+  Operation *getOperation() const { return operation; }
+
+  std::optional<MemoryAccessMatrix>
+  getMemoryAccessMatrix(const affine::MemRefAccess &access) const;
+
+private:
+  /// Construct the access map for the operation associated with the
+  /// analysis.
+  void build();
+
+  /// Attempt tp construct the access map entry for the given memory
+  /// operation \p memoryOp.
+  template <typename T> void build(T memoryOp, DataFlowSolver &solver);
+
+  /// Returns true if the memory access \p access has a single subscript that is
+  /// zero, and false otherwise.
+  bool hasZeroIndex(const affine::MemRefAccess &access) const;
+
+  /// Returns the unique definition for the operand at index \p opIndex in
+  /// operation \p op, or std::nullopt if it does not have a unique definition.
+  std::optional<Definition> getUniqueDefinition(unsigned opIndex, Operation *op,
```

Comment:
would `getUniqueModifier` be a better name?

Reply:
Why not directly take the Value?
```suggestion
  std::optional<Definition> getUniqueModifier(Value val,
```